### PR TITLE
Add Experimental Models for Apple M1 Firestorm and Icestorm

### PR DESCRIPTION
**Pull request by dop-amin**
This PR adds experimental models for Apple M1 Firestorm and Icestorm based on [work](https://dougallj.github.io/applecpu/firestorm.html) by [Dougall Johnson](https://twitter.com/dougallj). 
It also includes some optimized Dilithium NTT code.

Todos:

- [x] Add optimized Kyber NTTs for Firestorm
- [x] Add optimized Kyber and Dilithium NTTs for Icestorm
- [x] Extend `example.py` accordingly
- [x] Add all instructions that we support for A55 to the models

### CODE REVIEW

**Comment by hanno-becker** [comment]
@dop-amin Do I understand correctly that you aim to finish the TODOs before the PR should be merged?

**Comment by dop-amin** [comment]
> @dop-amin Do I understand correctly that you aim to finish the TODOs before the PR should be merged?

Yes, that's the plan!

**Comment by mkannwischer** [comment]
Maybe you want to switch to a draft PR then. Thanks for all the work!

**Comment by dop-amin** [comment]
I think the work on this PR is done.
There are some nuances to the M1 {Fire,Ice}storm models that are not captured yet but they have not been necessary for now. For quite some instructions, I also needed to do some guessing on the execution units/latencies; I hope this is somewhat reasonable. 

[A PR](https://github.com/slothy-optimizer/pqax/pull/2) that adds the tests for M1 has been made to [pqax](https://github.com/slothy-optimizer/pqax) aswell. 

NTT performance results, benchmarked on an M3, can be found below.

## Kyber
### M3 Firestorm
```
bench ntt_kyber ntt_kyber_123_4567                                 242 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load                     582 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store               534 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store                    265 cycles 1000 repeats
bench ntt_kyber ntt_kyber_1234_567                                 246 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_manual_st4_opt_a55              235 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_opt_a55                         228 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_opt_a55             368 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store_opt_a55       415 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store_opt_a55            242 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_manual_st4_opt_a72              230 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_opt_a72                         226 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_opt_a72             361 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store_opt_a72       383 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store_opt_a72            236 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_opt_m1_firestorm                228 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_opt_m1_firestorm    348 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store_opt_m1_firestorm 389 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_manual_st4_opt_m1_firestorm     228 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store_opt_m1_firestorm   241 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_manual_st4_opt_m1_icestorm      227 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_opt_m1_icestorm                 225 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_opt_m1_icestorm     412 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store_opt_m1_icestorm 401 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store_opt_m1_icestorm    238 cycles 1000 repeats
bench ntt_kyber ntt                                                260 cycles 1000 repeats
bench ntt_kyber pqclean_ntt                                        222 cycles 1000 repeats
```
### M3 Icestorm
```
bench ntt_kyber ntt_kyber_123_4567                                 527 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load                     695 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store               713 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store                    553 cycles 1000 repeats
bench ntt_kyber ntt_kyber_1234_567                                 481 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_manual_st4_opt_a55              448 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_opt_a55                         435 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_opt_a55             546 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store_opt_a55       570 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store_opt_a55            470 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_manual_st4_opt_a72              435 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_opt_a72                         426 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_opt_a72             562 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store_opt_a72       592 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store_opt_a72            449 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_opt_m1_firestorm                423 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_opt_m1_firestorm    544 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store_opt_m1_firestorm 591 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_manual_st4_opt_m1_firestorm     421 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store_opt_m1_firestorm   448 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_manual_st4_opt_m1_icestorm      423 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_opt_m1_icestorm                 425 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_opt_m1_icestorm     579 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_load_store_opt_m1_icestorm 626 cycles 1000 repeats
bench ntt_kyber ntt_kyber_123_4567_scalar_store_opt_m1_icestorm    462 cycles 1000 repeats
bench ntt_kyber ntt                                                518 cycles 1000 repeats
bench ntt_kyber pqclean_ntt                                        420 cycles 1000 repeats
```
---
## Dilithium
### M3 Firestorm
```
bench ntt_dilithium ntt_dilithium_123_45678                            429 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_w_scalar                   758 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4                 427 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678                            433 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_manual_st4                 432 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_opt_a55                    406 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4_opt_a55         415 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_w_scalar_opt_a55           629 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_opt_a72                    413 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4_opt_a72         413 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_opt_a72                    416 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_opt_m1_firestorm           405 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4_opt_m1_firestorm 405 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_opt_m1_firestorm           414 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_manual_st4_opt_m1_firestorm 410 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_opt_m1_icestorm            416 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4_opt_m1_icestorm 410 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_w_scalar_opt_m1_icestorm   652 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_opt_m1_icestorm            417 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_manual_st4_opt_m1_icestorm 413 cycles 1000 repeats
bench ntt_dilithium ntt                                                471 cycles 1000 repeats
bench ntt_dilithium pqclean_ntt                                        414 cycles 1000 repeats
```
### M3 Icestorm
```
bench ntt_dilithium ntt_dilithium_123_45678                            858 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_w_scalar                  1255 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4                 866 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678                            959 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_manual_st4                 969 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_opt_a55                    822 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4_opt_a55         845 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_w_scalar_opt_a55          1062 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_opt_a72                    814 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4_opt_a72         810 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_opt_a72                    824 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_opt_m1_firestorm           790 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4_opt_m1_firestorm 787 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_opt_m1_firestorm           898 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_manual_st4_opt_m1_firestorm 811 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_opt_m1_icestorm            788 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_manual_st4_opt_m1_icestorm 787 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_123_45678_w_scalar_opt_m1_icestorm  1108 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_opt_m1_icestorm            890 cycles 1000 repeats
bench ntt_dilithium ntt_dilithium_1234_5678_manual_st4_opt_m1_icestorm 821 cycles 1000 repeats
bench ntt_dilithium ntt                                                972 cycles 1000 repeats
bench ntt_dilithium pqclean_ntt                                        815 cycles 1000 repeats
```

File: example.py [hunk 1]
```diff
@@ -894,24 +984,34 @@ def __init__(self, var="", arch=AArch64_Neon, target=Target_CortexA72):
...
+                f"x{i}" for i in range(0, 6)] + ["x30", "sp"]
+            slothy.config.reserved_regs += self.target_reserved
+            slothy.config.inputs_are_outputs = True
+            slothy.config.sw_pipelining.halving_heuristic = True
+            slothy.config.split_heuristic = True
+            slothy.config.split_heuristic_factor = 2
+            slothy.config.split_heuristic_repeat = 4
+            slothy.config.split_heuristic_stepsize = 0.1
+            slothy.config.constraints.stalls_first_attempt = 14
+            # slothy.optimize_loop("layer1234_start")
```
Comment (hanno-becker) [inline example.py:1002]:
```suggestion
            slothy.optimize_loop("layer1234_start")
```

File: example.py [hunk 2]
```diff
@@ -529,15 +581,21 @@ def core(self,slothy):
 
         # layer567 is small enough for SW pipelining without heuristics
         slothy.config = Config(self.arch, self.target)
+        slothy.config.timeout = self.timeout
+        # Increase the timeout when not using heuristics
+        slothy.config.timeout = self.timeout * 12
```
Comment (hanno-becker) [inline example.py:586]:
```suggestion
        if self.timeout is not None:
                slothy.config.timeout = self.timeout * 12
```

**Comment by hanno-becker** [comment]
Thanks a lot @dop-amin, looks good.

**Pull request merged**